### DISCUSSION TITLE
yolov5/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#8575] [Docs]
# yolov5/

Deep dive into Ultralytics' YOLOv5. Learn about object detection model - YOLOv5, how to train it on custom data, multi-GPU training and more.

https://docs.ultralytics.com/yolov5/

**Comment by ajay78123** [comment]
I have trained yolov5 on custom dataset and gave that code and best.pt to backend team in the first phase its working but after one month i have tuned the model and send that best.pt from that ppoint after chnaging best.pt we are getting an error 

(torch) ubuntu@ip-172-31-19-90:~$ python3 torch_yolo/app.py 
w== /home/ubuntu/ModelWeights/SM/best.pt /home/ubuntu/ModelWeights/SM/best.pt
Traceback (most recent call last):
  File "/home/ubuntu/torch_yolo/app.py", line 14, in <module>
    predictor=YOLOv5Detector(SM)
  File "/home/ubuntu/torch_yolo/predictions.py", line 12, in _init_
    self.model = attempt_load(model_weights, device=self.device)
  File "/home/ubuntu/torch_yolo/models/experimental.py", line 80, in attempt_load
    ckpt = torch.load(attempt_download(w), map_location='cpu')  # load
  File "/home/ubuntu/miniconda3/envs/torch/lib/python3.9/site-packages/torch/serialization.py", line 1014, in load
    return _load(opened_zipfile,
  File "/home/ubuntu/miniconda3/envs/torch/lib/python3.9/site-packages/torch/serialization.py", line 1422, in _load
    result = unpickler.load()
  File "/home/ubuntu/miniconda3/envs/torch/lib/python3.9/pathlib.py", line 1084, in _new_
    raise NotImplementedError("cannot instantiate %r on your system"
NotImplementedError: cannot instantiate 'WindowsPath' on your system

in my windows machine it is working fine but it is causing an issue there
i have also tried to clone latest yolov5 repo and intstall the dependencies  but this is not resolving
if you have suggestions for us pleaase share

**Reply by glenn-jocher** [reply]
Hey there! 👋 It looks like you're encountering an issue with the `NotImplementedError: cannot instantiate 'WindowsPath' on your system` error when trying to load your model on an Ubuntu system, despite it working fine on Windows.

This error typically occurs due to path differences between Windows and Unix-like systems (like Ubuntu). The model path is being interpreted as a Windows path (`WindowsPath`) which isn't compatible with Unix-like systems.

A quick fix you might try is ensuring the path is correctly formatted as a string when loading your model. For example:

```python
# Instead of directly using the path object, convert it to a string
model_weights = str("/home/ubuntu/ModelWeights/SM/best.pt")
predictor = YOLOv5Detector(model_weights)
```

This ensures Python treats the path as a string rather than a `Path` object, which should be system-agnostic.

Also, make sure that when you're transferring the `.pt` file between Windows and Ubuntu, the path doesn't get formatted into a Windows-specific format. It's a good practice to define paths as raw strings or use forward slashes (`/`) and avoid backslashes (`\`) which are typical in Windows paths.

If you've already tried cloning the latest YOLOv5 repo and the issue persists, this path formatting might be the culprit. Give this a shot and let us know how it goes! 🚀

**Comment by ajay78123** [comment]
Thankyou for your valuable suggestion the is resolved.
Thanks for the support

**Reply by glenn-jocher** [reply]
Hey there! 🚀

I'm thrilled to hear the issue is resolved, and I appreciate your kind words and support. It's always great to see the community coming together to tackle challenges. If you have any more questions or run into new issues, feel free to reach out. We're here to help!

Happy coding! 😊

**Comment by ajay78123** [comment]
Athe issue is resolved

**Reply by pderrenger** [reply]
Great to hear the issue is resolved! 🎉 If you have any more questions or run into other issues, feel free to reach out. Happy coding with YOLOv8!

**Comment by DARVESH001** [comment]
how i can use my trained weights "best.pt". using opencv

**Reply by pderrenger** [reply]
Hi there! 😊

To use your trained weights `best.pt` with OpenCV, you can convert the model to ONNX format first, as OpenCV has good support for ONNX models. Here's how you can do it:

1. **Export to ONNX:**
```bash
yolo export model=best.pt format=onnx
```

2. **Use the ONNX model in OpenCV:**
```python
import cv2
import numpy as np

# Load the ONNX model
net = cv2.dnn.readNetFromONNX('best.onnx')

# Read an image
img = cv2.imread('your_image.jpg')

# Preprocess the image: resize, convert to blob
blob = cv2.dnn.blobFromImage(img, scalefactor=1/255, size=(640, 640), mean=(0,0,0), swapRB=True, crop=False)
net.setInput(blob)

# Run inference
outs = net.forward()

# Process the outputs as per your requirement...

# Display the image
cv2.imshow('Image', img)
cv2.waitKey(0)
cv2.destroyAllWindows()
```

Make sure to adjust the size parameter in `blobFromImage` according to your model's input size. Hope this helps get you started using your `best.pt` with OpenCV! 🚀

For any more questions, feel free to ask. Happy coding!

**Comment by mature-1111** [comment]
Traceback (most recent call last):
  File "D:/zyqi/yolov9-main/segment/train.py", line 646, in <module>
    main(opt)
  File "D:/zyqi/yolov9-main/segment/train.py", line 542, in main
    train(opt.hyp, opt, device, callbacks)
  File "D:/zyqi/yolov9-main/segment/train.py", line 270, in train
    for i, (imgs, targets, paths, _, masks) in pbar:  # batch ------------------------------------------------------
  File "D:\anaconda3\envs\yolov9\lib\site-packages\tqdm\std.py", line 1181, in __iter__
    for obj in iterable:
  File "D:\zyqi\yolov9-main\utils\dataloaders.py", line 170, in __iter__
    yield next(self.iterator)
  File "D:\anaconda3\envs\yolov9\lib\site-packages\torch\utils\data\dataloader.py", line 628, in __next__
    data = self._next_data()
  File "D:\anaconda3\envs\yolov9\lib\site-packages\torch\utils\data\dataloader.py", line 671, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
  File "D:\anaconda3\envs\yolov9\lib\site-packages\torch\utils\data\_utils\fetch.py", line 58, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "D:\anaconda3\envs\yolov9\lib\site-packages\torch\utils\data\_utils\fetch.py", line 58, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "D:\zyqi\yolov9-main\utils\segment\dataloaders.py", line 116, in __getitem__
    img, labels, segments = mixup(img, labels, segments, *self.load_mosaic(random.randint(0, self.n - 1)))
  File "D:\zyqi\yolov9-main\utils\segment\augmentations.py", line 16, in mixup
    segments = np.concatenate((segments, segments2), 0)
  File "<__array_function__ internals>", line 200, in concatenate
ValueError: all the input arrays must have same number of dimensions, but the array at index 0 has 1 dimension(s) and the array at index 1 has 3 dimension(s)

**Reply by mature-1111** [reply]
The problem   appear in yolov9 Instance Segmentation

**Reply by glenn-jocher** [reply]
Hi there 👋! It seems like you're encountering an issue with dimension mismatch during the mixup augmentation process in YOLOv9's instance segmentation. This error typically occurs when the arrays you're trying to concatenate don't have the same shape — in this case, a discrepancy between the number of dimensions.

Without seeing the exact code or data structure you're working with, it's a bit difficult to provide a precise solution. However, one common approach is to ensure that both arrays being concatenated (segments and segments2) have the same dimensionality before performing the operation. For instance, if `segments` is a 1D array and `segments2` is 3D, you might need to reshape or pad `segments` to match the dimensionality of `segments2`.

Here's a rough example of how to ensure they have the same number of dimensions:
```python
import numpy as np

# Assuming segments is 1D and segments2 is 3D
# Reshape or pad segments to make its dimensions compatible with segments2
segments = np.expand_dims(segments, axis=(1, 2))  # Adjust this based on your specific shapes

# Now that both have the same number of dimensions, concatenate should work
segments_combined = np.concatenate((segments, segments2), 0)
```

Remember, the key here is to adjust the dimensions so they're compatible. The exact method (`expand_dims`, `reshape`, `pad`, etc.) will depend on your data.

If you continue to face issues, could you provide more details or a snippet of the code where this error occurs? That way, we can offer more targeted advice. Happy coding! 🚀

**Comment by mature-1111** [comment]
Yes,I do it. I have converted from json format with Polygon labels to YOLOv7 instance segmentation. The solution to this problem is to&nbsp; use hyp.scratch-high.yaml but set mixup to 0.0
&nbsp;


天空
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Glenn ***@***.***&gt;; 
发送时间: 2024年4月18日(星期四) 上午6:56
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [ultralytics/ultralytics] yolov5/ (Discussion #8575)














 
Hi there 👋! It seems like you're encountering an issue with dimension mismatch during the mixup augmentation process in YOLOv9's instance segmentation. This error typically occurs when the arrays you're trying to concatenate don't have the same shape — in this case, a discrepancy between the number of dimensions.
 
Without seeing the exact code or data structure you're working with, it's a bit difficult to provide a precise solution. However, one common approach is to ensure that both arrays being concatenated (segments and segments2) have the same dimensionality before performing the operation. For instance, if segments is a 1D array and segments2 is 3D, you might need to reshape or pad segments to match the dimensionality of segments2.
 
Here's a rough example of how to ensure they have the same number of dimensions:
 import numpy as np # Assuming segments is 1D and segments2 is 3D # Reshape or pad segments to make its dimensions compatible with segments2 segments = np.expand_dims(segments, axis=(1, 2))  # Adjust this based on your specific shapes # Now that both have the same number of dimensions, concatenate should work segments_combined = np.concatenate((segments, segments2), 0)
 
Remember, the key here is to adjust the dimensions so they're compatible. The exact method (expand_dims, reshape, pad, etc.) will depend on your data.
 
If you continue to face issues, could you provide more details or a snippet of the code where this error occurs? That way, we can offer more targeted advice. Happy coding! 🚀

**Reply by glenn-jocher** [reply]
Hey there! 😊 It's great to hear you've successfully transitioned from JSON format with Polygon labels to YOLOv7 instance segmentation using the `hyp.scratch-high.yaml` and tweaking mixup to `0.0`. It sounds like you encountered a similar dimensionality issue during augmentation in YOLOv9 as discussed previously.

For dimension mismatch errors, ensuring that both arrays to be concatenated have matching shapes is indeed crucial. The previous suggestion to harmonize dimensions using operations like `np.expand_dims` or `reshape` before concatenation is a solid approach 👍.

Your initiative to share your solution and actively participate in this discussion enriches the community. If you hit any other snags or have further insights, feel free to share. Happy coding! 🚀

**Comment by nickguim** [comment]
Is there any tutorials on utilizing an object tracker such as deepsort or sort with an older model like yolov5? Everytime I attempt to use a github I have found something breaks

**Reply by pderrenger** [reply]
Hi there! 👋 It seems like you're looking to use an object tracker like deepsort or sort with YOLOv5 and encountering some issues. While specific tutorials on integrating these trackers with older models like YOLOv5 might not be readily available on our documentation, I suggest checking out the YOLOv5 GitHub repo or community discussions for any shared solutions or advice from other users who might have attempted something similar.

Integration usually involves passing YOLOv5 detection outputs as inputs to the tracker. Here's a very basic snippet to give you an idea of how you might start:

```python
# Assuming 'detections' is a list of detections from YOLOv5, 
# and 'tracker' is an instance of your tracking algorithm (e.g., deepsort)

for detection in detections:
    bbox, confidence, class_id = detection
    # Convert bbox from x1y1x2y2 to xywh format expected by some trackers
    xywh_bbox = convert_to_xywh(bbox) 
    # Now, you can update your tracker per frame
    tracker.update(xywh_bbox, confidence, class_id)
```

Unfortunately, integrating third-party trackers can sometimes lead to compatibility issues due to updates or differences in expected inputs/outputs. If specifics break, consider checking:

1. The version compatibility between YOLOv5 and the tracking library.
2. The format of the detection outputs and expected tracker inputs.
3. Any recent changes or updates in the tracker's source code that might affect integration.

For troubleshooting, I'd recommend revisiting the tracker documentation or seeking help directly from the tracker's GitHub page. Sometimes, the community might have already provided solutions for similar issues. 🛠️

Let us know if you encounter any specific errors or need further guidance! Our community is always here to help. Good luck with your project!

**Comment by Sequential-circuits** [comment]
Trying it in an nvidia jetson nano with jetpack 4.6 installed and getting AttributeError: partially initialized module 'cv2' has no attribute '_registerMatType' (most likely due to a circular import)

**Reply by pderrenger** [reply]
@Sequential-circuits hey there! It looks like you're facing an `AttributeError` related to OpenCV on your NVIDIA Jetson Nano. This issue might indeed be tied to a circular import or an incompatible OpenCV version. 😕

Here's a quick suggestion:
1. Ensure OpenCV is correctly installed and updated. Sometimes, reinstalling can help resolve strange errors. You can reinstall OpenCV using:
   ```bash
   pip install opencv-python-headless --force-reinstall
   ```
   Given the NVIDIA platform, ensure all dependencies and JetPack SDK components are compatible and up-to-date.

If the issue persists, consider checking the OpenCV installation specific to Jetson devices or exploring the NVIDIA forums for any JetPack-related quirks. The Jetson community is quite active and may have encountered similar issues.

Wishing you smoother sailing ahead! 🚀 If you have more questions or updates, feel free to reach out.

**Comment by ghitabnr** [comment]
hello please i trained the yolov5 in my custom dataset that contains two classes but when i do the test the two classes show up with very close colors ( red and pink ) i want to the change the color of the boxes of detection? how please ?

**Reply by glenn-jocher** [reply]
@ghitabnr hi there! 👋 Changing the color of detection boxes in YOLOv5 involves modifying the `plot_one_box()` function typically found in the `utils/plots.py` file. You'll have the flexibility to assign different colors to different classes.

Here's a quick example of how you can modify it:

1. Locate the `plot_one_box()` function.
2. Inside this function, you'll find a line that sets the color. Modify this line to choose your colors based on class IDs. For instance:
   ```python
   color = colors[int(cls)]  # default color selection
   ```

   Change to:
   ```python
   colors = [[0,255,0], [255,0,0]]  # Define colors for each class, e.g., green and blue
   color = colors[int(cls)]
   ```

Remember to define enough colors for each class you have to distinguish them clearly.

Hope this helps! Let me know if there's anything else you'd like to tweak! 🚀

**Comment by ShashankMPatil** [comment]
Why i am not able to open webcam

import cv2
import numpy as np
import torch

# my best.pt is in content folder

model = torch.hub.load('ultralytics/yolov5', 'custom', path='/content/best.pt', force_reload=True)

cap =cv2.VideoCapture(2)
while cap.isOpened():
  ret,frame=cap.read()

  results=model(frame)

  cv2.imshow("YOLO", np.squeeze(results.render()))



even if i enter below code , webcam doesn't open
!python detect.py --source 0

  if cv2.waitKey(10) & 0xFF == ord('q'):
    break

cap.release()
cv2.destroyAllWindows()

**Reply by ShashankMPatil** [reply]
the code is bit jumbled, i hope you get what i am asking

**Reply by glenn-jocher** [reply]
It looks like you're having trouble with opening the webcam using OpenCV and YOLOv5. Here are a few things you might want to check:

1. **Webcam Index**: Ensure that the webcam index you are using (`2` in your case) is correct. Typically, the default webcam is at index `0`. If you have multiple cameras, you might need to try different indices like `0` or `1`.

2. **Camera Permissions**: Make sure that your system has the necessary permissions to access the webcam.

3. **Code Check**: Ensure that your code for opening the webcam and reading frames is correct. Here's a simplified version to test webcam functionality:

   ```python
   import cv2

   cap = cv2.VideoCapture(0)  # Try changing this index to 0 or 1
   if not cap.isOpened():
       print("Cannot open camera")
       exit()
   while True:
       ret, frame = cap.read()
       if not ret:
           print("Can't receive frame (stream end?). Exiting ...")
           break
       cv2.imshow('frame', frame)
       if cv2.waitKey(1) == ord('q'):
           break
   cap.release()
   cv2.destroyAllWindows()
   ```

4. **Driver Issues**: Sometimes, driver issues can prevent access to the webcam. Ensure your device drivers are up to date.

If you're still facing issues, it might be helpful to provide more details or check the camera with other software to isolate the problem.

**Comment by roihanmjd** [comment]
i have a problem like this :
(base) D:\yolov5update\yolov5>python detect.py
Traceback (most recent call last):
  File "D:\yolov5update\yolov5\detect.py", line 38, in <module>
    import torch
  File "C:\Users\ASUS\AppData\Roaming\Python\Python311\site-packages\torch\__init__.py", line 141, in <module>
    raise err
OSError: [WinError 126] The specified module could not be found. Error loading "C:\Users\ASUS\AppData\Roaming\Python\Python311\site-packages\torch\lib\shm.dll" or one of its dependencies.

**Reply by pderrenger** [reply]
It looks like there's an issue with loading the PyTorch library due to a missing `shm.dll` file. This can happen if the PyTorch installation is incomplete or corrupted. Here are a few steps to resolve this:

1. **Reinstall PyTorch**: Ensure you have the correct version of PyTorch installed for your Python version. You can reinstall it using pip:

   ```bash
   pip install torch --force-reinstall
   ```

2. **Check System Path**: Ensure that the path to your Python environment is correctly set in your system's PATH variable.

3. **Dependencies**: Sometimes, missing system dependencies can cause this error. You might want to ensure all necessary system dependencies for PyTorch are installed.

If the issue persists after trying these steps, consider reinstalling Python and PyTorch in a new virtual environment. This can sometimes resolve hidden conflicts or corrupted installations.

**Reply by dalmeidalorraine** [reply]
Does YOLO v5 support multi label classification? If yes, where can I find some examples and documentation?

**Reply by pderrenger** [reply]
@dalmeidalorraine yes, YOLOv5 supports multi-label classification! You can train a model to detect multiple labels per detection by adjusting your dataset annotations to reflect multiple classes per bounding box. For examples and more detailed instructions on setting this up, please refer to the "Train Custom Data" section of our comprehensive YOLOv5 documentation. This section will guide you through the process of training YOLOv5 on your custom dataset, which can include multi-label scenarios. Happy modeling! 🚀

**Comment by dalmeidalorraine** [comment]
Does YOLO v5 support multi label classification? If yes, where can I find some examples and documentation?

**Reply by pderrenger** [reply]
Yes, YOLOv5 does support multi-label classification. You can find examples and detailed documentation on how to train YOLOv5 on your custom dataset, which can include multi-label scenarios, in our comprehensive guide. Please refer to the "Train Custom Data" section of the YOLOv5 documentation for step-by-step instructions and tips to optimize your training process. This will help you effectively handle multiple labels per image in your projects. Happy modeling! 🚀

**Comment by Vidhul-S** [comment]
Hello i am new to this but her i go,
i am supposed to create a streamlined way to run a yolov5 model on a jetson nano
since it runs runs ubuntu 18.04 and torch limitations i decided to go with virtual environment 
But i am still limited to python 3.6.9 because the last compiled torch available 1.10.0 only works on 3.6.9 or lower
So i had to use an older version of yolov5 namely Yolov5 v6.1 on the nano
Now i am training my models on a much better system and only inferencing on the nano 
My issue starts here 
First i Tried to train a model on yolov5 v7.0 (latest one)
and tried to inference in on nano and obviously it didn't run 
so i tried to checkout v6.1 on my system but i am un able to get it running due to the dependencies 
I tried everything i can think off ranging from using different 
python version and environments to modifying the requirements.txt 
to forcefully chance the versions ,I couldn't get a working project which can train models 
can anyone provide me a working set of versions of all these packages to install .
Or link to a project with yolov5 v6.1


Note: 
>The machine i am training models on is a really capable one so no issue in that front
>I am not a programmer and still learning so pardon me for anything wrong above
>English is not my first language so if anything is confusing firstly i am sorry feel free to ask for any clarification

**Reply by Vidhul-S** [reply]
let me know if there is any other way i can get yolov5 running on jetson nano with pytorch 1.10.0 python3.6.9

**Reply by glenn-jocher** [reply]
@Vidhul-S running YOLOv5 on a Jetson Nano with PyTorch 1.10.0 and Python 3.6.9 can be challenging due to compatibility issues. Given your constraints, using YOLOv5 v6.1 is a reasonable approach. To streamline your workflow, ensure you match the dependencies of YOLOv5 v6.1 on both your training system and the Jetson Nano. You can find the specific versions of dependencies in the `requirements.txt` file of the YOLOv5 v6.1 repository. If you encounter issues, consider using Docker to create a consistent environment across different systems. For further guidance, refer to the [YOLOv5 documentation](https://docs.ultralytics.com/yolov5/).

**Comment by sustaintipsdeveloper** [comment]
Hello, I wanted to calculate the accuracy of the YOLOv5. I'm doing a physical testing. Can you recommend me, what is the formula that can I used for this physical testing? Is the formula will be like this?
Accuracy = Sum of Confidence Score/Total Test Cases × 100

**Reply by glenn-jocher** [reply]
@sustaintipsdeveloper the formula you mentioned, `Accuracy = Sum of Confidence Score / Total Test Cases × 100`, is not typically used to calculate accuracy in object detection. Instead, metrics like Mean Average Precision (mAP) and Intersection over Union (IoU) are commonly used. You can refer to the [YOLOv5 documentation](https://docs.ultralytics.com/yolov5/) for guidance on evaluating model performance using these metrics. Let me know if you need further clarification!

**Reply by sustaintipsdeveloper** [reply]
Sir, how about the confidence level? What's the formula? Thank you

**Reply by glenn-jocher** [reply]
The confidence level in YOLO models is the score that represents the model's certainty that a detected object belongs to a particular class. It is not calculated manually but is an output of the model, ranging from 0 to 1. You can adjust the confidence threshold during inference to filter predictions. For more details on confidence scores, refer to the [YOLOv5 documentation](https://docs.ultralytics.com/yolov5/). Let me know if you need further clarification!
